signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt

```diff
@@ -0,0 +1,214 @@
+package io.provenance.hdwallet.signer
+
+import io.provenance.hdwallet.ec.Curve
+import java.io.ByteArrayOutputStream
+import java.math.BigInteger
+import org.bouncycastle.asn1.ASN1InputStream
+import java.security.Signature as JavaSignature
+import org.bouncycastle.asn1.ASN1Integer
+import org.bouncycastle.asn1.ASN1Primitive
+import org.bouncycastle.asn1.ASN1Sequence
+import org.bouncycastle.asn1.DERSequenceGenerator
+
+@JvmInline
+value class BTCSignature private constructor(private val value: ByteArray) {
+    companion object {
+        fun fromByteArray(bytes: ByteArray): BTCSignature = BTCSignature(bytes)
+    }
+
+    fun toByteArray(): ByteArray = value
+}
+
+@JvmInline
+value class ASN1Signature private constructor(private val value: ByteArray) {
+    companion object {
+        fun fromByteArray(bytes: ByteArray): ASN1Signature = ASN1Signature(bytes)
+    }
+
+    fun toByteArray(): ByteArray = value
+    fun toIntegerPair(): BigIntegerPair = ASN1.decode(this)
+}
+
+@JvmInline
+value class BigIntegerPair(private val intPair: Pair<BigInteger, BigInteger>) {
+    val r get(): BigInteger = intPair.first
+    val s get(): BigInteger = intPair.second
+    fun asPair(): Pair<BigInteger, BigInteger> = intPair
+}
+
+/**
+ * Bitcoin-specific signature format operations.
+ */
+object BTC {
+    /**
+     * [encode] returns the ECDSA signature as a ByteArray of `r || s`, * where both `r` and `s` are encoded into 32
+     * byte big endian integers.
+     */
+    fun encode(r: BigInteger, s: BigInteger, curve: Curve): BTCSignature {
+
+        val halfCurveOrder = curve.n.shiftRight(1)
+
+        /**
+         * Returns the bytes from a BigInteger as an unsigned version by truncating a byte if needed.
+         *
+         * @return `r || s` encoded as an array of bytes.
+         */
+        fun BigInteger.getUnsignedBytes(): ByteArray {
+            val bytes = this.toByteArray()
+            if (bytes[0] == 0x0.toByte()) {
+                return bytes.drop(1).toByteArray()
+            }
+            return bytes
+        }
+
+        // Canonicalize - In order to remove malleability,
+        // we set s = curve_order - s, if s is greater than curve.Order() / 2.
+        var sigS = s
+        if (sigS > halfCurveOrder) {
+            sigS = curve.n.subtract(sigS)
+        }
+
+        val sBytes = sigS.getUnsignedBytes()
+        val rBytes = r.getUnsignedBytes()
+
+        require(rBytes.size <= 32) { "cannot encode r into BTC Format, size overflow (${rBytes.size} > 32)" }
+        require(sBytes.size <= 32) { "cannot encode s into BTC Format, size overflow (${sBytes.size} > 32)" }
+
+        val signature = ByteArray(64)
+        // 0 pad the byte arrays from the left if they aren't big enough.
+        System.arraycopy(rBytes, 0, signature, 32 - rBytes.size, rBytes.size)
+        System.arraycopy(sBytes, 0, signature, 64 - sBytes.size, sBytes.size)
+
+        return BTCSignature.fromByteArray(signature)
+    }
```

Comment:
Sibling object has both encode and decode, @mwoods-figure can we get the decode method pulled in here? I think that would be a little more consistent.

Reply:
@mtps Sure!